### PR TITLE
RHELPLAN-32453 - Add remote inputs (imtcp, imptcp, imudp) and their r…

### PR DESCRIPTION
**Pull request by nhosoi**
…emote_files outputs supports

- Separating the inputs from the remote hosts from the basic input type and
  making it a remote input type.
- Separating the output into the local file from the files output type and
  making it a remote_file output type, which uses an output_remote_files.j2
  template.

- CI tests
  . adding tests_remote_default_remote.yml for testing the default output rules.
  . adding tests_remote_remote.yml for testing the configured output rules.
  . removing an obsolete tests_listen.yml.

- Updating docs:
  . README.md
  . roles/rsyslog/README.md
  . design_docs/logging_subsystem.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -135,21 +135,46 @@ logging_inputs:
...
-    type: files
-logging_inputs:
-  - name: system-and-remote-input
-    type: basics
-logging_flows:
-  - name: flow0
-    inputs: [system-and-remote-input]
-    outputs: [local-files]
+        logging_outputs:
+          - name: default_remote_files
```
Comment (pcahyna) [inline README.md:141]: Why is it called `default_remote_files`? Is it a special name or just an arbitrary name that you introduced in the example? If it is just an arbitrary name, why is it called `default_something`? To me `default` sounds like something predefined.

File: README.md [hunk 2]
```diff
@@ -135,21 +135,46 @@ logging_inputs:
...
-logging_outputs:
-  - name: local-files
-    type: files
-logging_inputs:
-  - name: system-and-remote-input
-    type: basics
-logging_flows:
-  - name: flow0
-    inputs: [system-and-remote-input]
-    outputs: [local-files]
```
Comment (pcahyna) [inline README.md:152]: is there a reason to increase the indent of those lines?

File: tests/tests_remote_default_remote.yml [hunk 3]
```diff
@@ -27,6 +30,22 @@
...
+      assert:
+        that: rsyslog_conf_line_count.stdout.0 | int <= 7
+
+    - name: Check file counts in rsyslog.d
+      assert:
+        that: rsyslog_d_file_count.matched >= 8
+
+    - name: Check rsyslog errors
+      shell: |
+        systemctl status rsyslog | grep error
```
Comment (pcahyna) [inline tests/tests_remote_default_remote.yml:43]: please no `shell` and `grep`.

File: tests/tests_remote_default_remote.yml [hunk 4]
```diff
@@ -27,6 +30,22 @@
...
+      register: result
+      failed_when: result.rc == 0
+      tags:
+        - skip_ansible_lint
```
Comment (pcahyna) [inline tests/tests_remote_default_remote.yml:47]: what does this tag do?

File: README.md (hunk 1 above)
Comment (nhosoi) [reply README.md:141]: > Why is it called `default_remote_files`? Is it a special name or just an arbitrary name that you introduced in the example? If it is just an arbitrary name, why is it called `default_something`? To me `default` sounds like something predefined.

The name can be any string, but since there is no `remote_log_path` nor `remote_sub_path` is defined in the remote_files output, the [default configuration](https://github.com/linux-system-roles/logging/pull/116/files#diff-b23a18f65552498ad31bf927da164818R20-R55) is used.

File: README.md (hunk 2 above)
Comment (nhosoi) [reply README.md:152]: No, I don't think so...

File: README.md (hunk 1 above)
Comment (nhosoi) [reply README.md:141]: Replaced `default_remote_files` with `remote_files_output`.

File: roles/rsyslog/tasks/inputs/remote/main.yml [hunk 5]
```diff
@@ -9,15 +9,26 @@
...
   vars:
-    __rsyslog_input_decl: ""
-    __udp_input_calls: "# Log messages from remote hosts over UDP\ninput(name=\"{{ remote_item.name }}\" type=\"imudp\" port=\"{{ remote_item.udp_port | d(__rsyslog_default_port) }}\")\n"
-    __ptcp_input_calls: "# Log messages from remote hosts over plain TCP\ninput(name=\"{{ remote_item.name }}\" type=\"imptcp\" port=\"{{ remote_item.tcp_port | d(__rsyslog_default_port) }}\")\n"
-    __tls_input_calls: "# Log messages from remote hosts over TLS\ninput(name=\"{{ remote_item.name }}\" type=\"imtcp\" port=\"{{ remote_item.tcp_tls_port | d(__rsyslog_default_tls_port) }}\")\n"
     __rsyslog_input_name: "{{ remote_item.name }}"
     __rsyslog_input_type: "{{ remote_item.type }}"
     __rsyslog_packages: []
-    __rsyslog_rules: [ { "name": "input-remote-{{ remote_item.name }}", "type": "input", "weight": "11",
-                         "sections": [ { "options": "{{ __udp_input_calls if (remote_item.udp_port | d() ) else ( __ptcp_input_calls if remote_item.tcp_port | d() else __tls_input_calls ) }}{{ lookup('template', 'input_template.j2') }}" } ] } ]
```
Comment (pcahyna) [inline roles/rsyslog/tasks/inputs/remote/main.yml]: @nhosoi You say "__rsyslog_rules value from a string to the dictionary" but __rsyslog_rules was a dictionary before and it is a dictionary after the change.

Comment (nhosoi) [reply roles/rsyslog/tasks/inputs/remote/main.yml]: @pcahyna, indeed :p. I'll fix the comment to `modifying the __rsyslog_rules value to YAML friendly manner.`

File: README.md [hunk 6]
```diff
@@ -135,21 +135,46 @@ logging_inputs:
     type: basics
 ```
 
-3) Deploying basic LSR/Logging config files in /etc/rsyslog.d, which handle inputs from the local system and remote rsyslog and outputs into the local files.
+3) Deploying basic LSR/Logging config files in /etc/rsyslog.d, which handle inputs from the remote rsyslog and outputs into the per host base local files.
```
Comment (richm) [inline README.md:138]: did you mean "per host"?  not sure what "base" means in this context

File: README.md [hunk 7]
```diff
@@ -135,21 +135,46 @@ logging_inputs:
...
+    outputs: [remote_files_output]
+```
+If more detailed outputs to be configured instead of using the default paths, they are configurable as follows.
+```
+logging_outputs:
+  - name: remote_files_output0
+    type: remote_files
+    remote_log_path: /var/log/remote/%HOSTNAME%/%PROGRAMNAME:::secpath-replace%.log
+    severity: info
+    exclude: ;authpriv.none
```
Comment (richm) [inline README.md:162]: this will change if https://github.com/linux-system-roles/logging/pull/114 merges first

File: README.md [hunk 8]
```diff
@@ -235,10 +260,27 @@ Variables in vars.yml
       - `name`: Name of the custom file output element.
       - `type`: Type of the output element.
       - `custom_config_files`: List of custom configuration files are deployed to /etc/rsyslog.d. [ '/path/to/custom_A.conf', '/path/to/custom_B.conf' ]. Default to none.
+   -  ** `type: remote_files`**
+      - `facility`: facility; default to `*`
+      - `severity`: severity; default to `*`
+      - `exclude`: exclude; default to none.
```
Comment (richm) [inline README.md:266]: this will change too

File: README.md [hunk 9]
```diff
@@ -235,10 +260,27 @@ Variables in vars.yml
...
+   -  ** `type: basics`**
+      - `rsyslog_imjournal_ratelimit_burst`: set to imjournal RateLimit.Burst. Default to 20000.
+      - `rsyslog_imjournal_ratelimit_interval`: set to imjournal RateLimit.Interval. Default to 600.
+      - `rsyslog_imjournal_persist_state_interval`: set to imjournal PersistStateInterval. Default to 10.
+   -  ** `type: files`**
+      - `rsyslog_input_log_path`: File name to be read by the imfile plugin. The value should be full path. Wildcard '*' is allowed in the path.  Default to `/var/log/containers/*.log`.
+   -  ** `type: remote_files`**
+      - `udp_port`: if the port number is given, rsyslog is configured to listen at the udp port number. Default to 514.
+      - `tcp_port`: if the port number is given, rsyslog is configured to listen at the tcp port number. Default to 514.
+      - `tcp_tls_port`: if the port number is given, rsyslog is configured to listen at the tls tcp port number. Default to 6514.
```
Comment (richm) [inline README.md:283]: this needs certs, key?

File: tests/tests_remote_default_remote.yml [hunk 10]
```diff
@@ -27,6 +30,19 @@
...
+        that: rsyslog_conf_line_count.stdout.0 | int <= 7
+
+    - name: Check file counts in rsyslog.d
+      assert:
+        that: rsyslog_d_file_count.matched >= 8
+
+    - name: Check rsyslog errors
+      command: systemctl status rsyslog
+      register: systemctl_result
+      failed_when: "'error' in systemctl_result.stderr"
```
Comment (richm) [inline tests/tests_remote_default_remote.yml:44]: `or systemctl_result.rc != 0`

File: README.md (hunk 6 above)
Comment (nhosoi) [reply README.md:138]: Yes, I wanted to convert this line into a plain English... :p
`/var/log/remote/%HOSTNAME%/%PROGRAMNAME:::secpath-replace%.log`
With this path, remote logs are stored in the local files.
```
ls /var/log/remote/msg/_hostname_
abrt-hook-ccpp.log journal.log         newtag.log         setroubleshoot.log
auditd.log               kernel.log          nm-dispatcher.log  systemd.log
testTag0.log           ...
```
Is "outputs into the local files per host" better?

File: README.md (hunk 7 above)
Comment (nhosoi) [reply README.md:162]: Right. I'll fix the conflicts...

File: README.md (hunk 6 above)
Comment (richm) [reply README.md:138]: > Is "outputs into the local files per host" better?

Yes, works for me

File: README.md (hunk 9 above)
Comment (nhosoi) [reply README.md:283]: Right. The tls part hasn't been touched yet. :(

File: tests/tests_remote_default_remote.yml (hunk 10 above)
Comment (nhosoi) [reply tests/tests_remote_default_remote.yml:44]: Actually, the return value is 0...
Please see this comment and the commit in pr115.
https://github.com/linux-system-roles/logging/pull/115#issuecomment-627882173
https://github.com/linux-system-roles/logging/pull/115/files#diff-b66e58c3692b490cd622af3414c88ab1R75-R81

Comment (richm) [reply tests/tests_remote_default_remote.yml:44]: Right - I get that in some error cases, `rc` will be `0` and there will be the string `error` somewhere in the output.  But for some error cases, maybe not even ones that you are expecting, it is possible that `systemctl status rsyslog` will return an non-zero exit code, and the string `error` might not be in the output in that case (e.g. rsyslog crashed for some reason before being able to write out the full error message).  In that case, rsyslog will have failed, but you will have masked the error by using `failed_when`.

Comment (nhosoi) [reply tests/tests_remote_default_remote.yml]: Thanks, @richm. Fixed. Also, rebased with master.

File: roles/rsyslog/tasks/inputs/remote/main.yml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+              # Log messages from remote hosts over UDP
+              input(name="{{ remote_item.name }}" type="imudp" port="{{ remote_item.udp_port | d(__rsyslog_default_port) }}")
+              {% elif remote_item.tcp_port | d() %}
+              # Log messages from remote hosts over plain TCP
+              input(name="{{ remote_item.name }}" type="imptcp" port="{{ remote_item.tcp_port | d(__rsyslog_default_port) }}")
+              {% else %}
+              # Log messages from remote hosts over TLS
+              input(name="{{ remote_item.name }}" type="imtcp" port="{{ remote_item.tcp_tls_port | d(__rsyslog_default_tls_port) }}")
+              {% endif %}
+              {{ lookup('template', 'input_template.j2') }}
```
Comment (richm) [inline roles/rsyslog/tasks/inputs/remote/main.yml:31]: I suggest putting this into a template file `input_remote_template.j2` or something like that

File: roles/rsyslog/tasks/outputs/remote_files/main.yml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+---
+- name: Create files output configuration files in /etc/rsyslog.d
+  vars:
+    __rsyslog_packages: "{{ __rsyslog_remote_files_output_packages }}"
+    __rsyslog_rules: [ { "name": "output-files-{{ remote_files_item.name }}", "type": "output",
+                         "sections": [ { "options": "{{ lookup('template', 'output_remote_files.j2') }}" } ] } ]
```
Comment (richm) [inline roles/rsyslog/tasks/outputs/remote_files/main.yml:6]: I suggest cleaning this up to put it into a more yaml/ansible friendly structure

File: roles/rsyslog/tasks/inputs/remote/main.yml (hunk 11 above)
Comment (nhosoi) [reply roles/rsyslog/tasks/inputs/remote/main.yml:31]: Thanks, @richm. I'm updating the pr.

I was wondering which is preferable - sharing templates and making the number of templates as small as possible vs simplifying the task code. It sounds latter, doesn't it? It's not like an analogy like a template is a library function to be shared so that it's supposed to be generic. :)

Comment (richm) [reply roles/rsyslog/tasks/inputs/remote/main.yml]: > I was wondering which is preferable - sharing templates and making the number of templates as small as possible vs simplifying the task code. It sounds latter, doesn't it?

In this case, yes, but like anything, it is a judgement call - if the tasks code is "too complex", it should be moved into a template file, or otherwise refactored - for sure the yaml and ansible linters don't like really long lines, and trying to follow rules of indentation can sometimes reduce readability.  On the other hand, moving too much code to template files can also reduce readability if you have to constantly switch back and forth between the tasks and the template to find out what's going on.

File: roles/rsyslog/templates/input_remote.j2 [hunk 13]
```diff
@@ -0,0 +1,46 @@
+{% if remote_item.udp_port | d() %}
+# Log messages from remote hosts over UDP
+input(name="{{ remote_item.name }}" type="imudp" port="{{ remote_item.udp_port | d(__rsyslog_default_port) }}")
+{% elif remote_item.tcp_port | d() %}
+# Log messages from remote hosts over plain TCP
+input(name="{{ remote_item.name }}" type="imptcp" port="{{ remote_item.tcp_port | d(__rsyslog_default_port) }}")
+{% else %}
+# Log messages from remote hosts over TLS
+input(name="{{ remote_item.name }}" type="imtcp" port="{{ remote_item.tcp_tls_port | d(__rsyslog_default_tls_port) }}")
+{% endif %}
```
Comment (richm) [inline roles/rsyslog/templates/input_remote.j2:10]: I think you can use `{{ lookup('template', 'input_template.j2') }}` here to avoid having to copy/paste the contents of input_template.j2 here.  In that sense, yes, you could think of a template as a "library" of "functions" that you can "include" in multiple places.

https://docs.ansible.com/ansible/latest/plugins/lookup/template.html

Comment (nhosoi) [reply roles/rsyslog/templates/input_remote.j2:10]: Ah, I see. We can template in a template file... :O

File: tests/tests_remote_default_remote.yml (hunk 10 above)
Comment (pcahyna) [reply tests/tests_remote_default_remote.yml]: Indeed - `or` was part of the suggested code :-)

File: roles/rsyslog/tasks/inputs/remote/main.yml (hunk 11 above)
Comment (pcahyna) [reply roles/rsyslog/tasks/inputs/remote/main.yml]: looking up a template from another template gives you the best of both, doesn't it?

Comment (pcahyna) [reply roles/rsyslog/tasks/inputs/remote/main.yml]: Also, would template inheritance help as an alternative to the template lookup? https://jinja.palletsprojects.com/en/2.9.x/templates/#template-inheritance

File: roles/rsyslog/templates/input_remote.j2 (hunk 13 above)
Comment (pcahyna) [reply roles/rsyslog/templates/input_remote.j2:10]: Or inherit from templates, not sure whether it helps here though: https://jinja.palletsprojects.com/en/2.9.x/templates/#template-inheritance

File: roles/rsyslog/defaults/main.yml [hunk 14]
```diff
@@ -154,20 +154,15 @@ rsyslog_send_permitted_peers: '{{ rsyslog_permitted_peers }}'
 # Firewall, UDP, TCP ports
 # ------------------------
 
-# rsyslog_udp_port
+# default rsyslog port
 #
-# The incoming UDP port used for remote logging.
-rsyslog_udp_port: '514'
+# The incoming TCP/UDP port used for remote logging.
+__rsyslog_default_port: '514'
```
Comment (richm) [inline roles/rsyslog/defaults/main.yml:160]: private variables not settable by users should go into `vars/main.yml` - https://github.com/oasis-roles/meta_standards#vars-vs-defaults

File: roles/rsyslog/vars/inputs/remote/main.yml [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+# List of ``rsyslogd`` modules that enable logs from the local system to be
+# received and parsed by the ``rsyslogd`` daemon.
+__rsyslog_conf_remote_modules:
+
+  - name: 'input-remote-modules'
+    type: 'modules'
+    sections:
+
+      - options: |-
+          {% set ns0 = namespace(use_tls = false) %}
```
Comment (richm) [inline roles/rsyslog/vars/inputs/remote/main.yml:38]: This should be in a template file

**Comment by nhosoi** [comment]
Rebased against the master branch.

File: roles/rsyslog/defaults/main.yml (hunk 14 above)
Comment (nhosoi) [reply roles/rsyslog/defaults/main.yml:160]: This is a fallback port number used like this. Yes, I should move it to roles/rsyslog/vars/main.yml...
`input(name="{{ remote_item.name }}" type="imudp" port="{{ remote_item.udp_port | d(__rsyslog_default_port) }}")`

File: .travis.yml [hunk 16]
```diff
@@ -7,8 +7,8 @@ env:
   matrix:
     - MOLECULE_DISTRO="centos:7"
     - MOLECULE_DISTRO="centos:8"
-    - MOLECULE_DISTRO="fedora:30"
     - MOLECULE_DISTRO="fedora:31"
+    - MOLECULE_DISTRO="fedora:32"
```
Comment (richm) [inline .travis.yml:11]: please do this in a separate PR - unless this PR needs to specifically run the molecule tests on fedora 32

Comment (nhosoi) [reply .travis.yml:11]: Sure. Is it ok to remove fedora:30 in this PR? Otherwise, continuous-integration/travis-ci/pr does not pass...

**Comment by richm** [comment]
lgtm

**Comment by richm** [comment]
>  @nhosoi nhosoi force-pushed the nhosoi:32453 branch from cb3b9ae to 528565d 16 minutes ago 

What changed?

**Comment by nhosoi** [comment]
> > @nhosoi nhosoi force-pushed the nhosoi:32453 branch from [cb3b9ae](https://github.com/linux-system-roles/logging/commit/cb3b9aed48fc49db51e8ba75aa853b406886de38) to [528565d](https://github.com/linux-system-roles/logging/commit/528565da28bb570f1b00c229b26f1994b9774799) 16 minutes ago
> 
> What changed?

Squashed and cleaned up the github comment...

**Comment by richm** [comment]
> > > @nhosoi nhosoi force-pushed the nhosoi:32453 branch from [cb3b9ae](https://github.com/linux-system-roles/logging/commit/cb3b9aed48fc49db51e8ba75aa853b406886de38) to [528565d](https://github.com/linux-system-roles/logging/commit/528565da28bb570f1b00c229b26f1994b9774799) 16 minutes ago
> > 
> > 
> > What changed?
> 
> Squashed and cleaned up the github comment...
> I don't think it introduced the travis-ci failure, but somehow it failed on CentOS-8
> 
> ```
> TASK [/home/travis/build/linux-system-roles/logging/roles/rsyslog : Create a work directory] ***
> fatal: [instance]: FAILED! => {"changed": false, "module_stderr": "", "module_stdout": "OCI runtime exec failed: exec failed: container_linux.go:348: starting container process caused \"process_linux.go:90: adding pid 7145 to cgroups caused \\\"failed to write 7145 to cgroup.procs: write /sys/fs/cgroup/cpu,cpuacct/docker/a459ded0f0c7ec8fce720c5fdb33b490b89993d24f226aa11afa09fb7582682e/cgroup.procs: invalid argument\\\"\": unknown\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 126}
> ```

Can you reproduce locally?  It looks like it is failing on this task:
```yaml
    # Creating a work directory
    - name: Create a work directory
      file:
        state: directory
        path: '{{ rsyslog_work_dir }}'
```
where the default for `rsyslog_work_dir` is `/var/lib/rsyslog` - is it not allowed to create `/var/lib/rsyslog` in a centos8 container?

**Comment by nhosoi** [comment]
> Can you reproduce locally? It looks like it is failing on this task:
> 
> ```yaml
>     # Creating a work directory
>     - name: Create a work directory
>       file:
>         state: directory
>         path: '{{ rsyslog_work_dir }}'
> ```
> 
> where the default for `rsyslog_work_dir` is `/var/lib/rsyslog` - is it not allowed to create `/var/lib/rsyslog` in a centos8 container?

Well, the path is taken in every CI-test and there's no failure observed. Not sure why, but I feel the travis-ci randomly fails in these couple of days...

**Comment by richm** [comment]
> Well, the path is taken in every CI-test and there's no failure observed. Not sure why, but I feel the travis-ci randomly fails in these couple of days...

Ok.  Not sure what's going on.  Also, I'm not aware of a way to tell travis-ci to re-run a failed test.  If you can't figure out a way to do this, then submit a "dummy" commit i.e. make a very minor change to the README or something like that, to trigger a new travis-ci run.

**Comment by nhosoi** [comment]
> > Well, the path is taken in every CI-test and there's no failure observed. Not sure why, but I feel the travis-ci randomly fails in these couple of days...
> 
> Ok. Not sure what's going on. Also, I'm not aware of a way to tell travis-ci to re-run a failed test. If you can't figure out a way to do this, then submit a "dummy" commit i.e. make a very minor change to the README or something like that, to trigger a new travis-ci run.

I changed one character in the github comment and forced to push :). Glad the travis-ci passed this time.

**Comment by nhosoi** [comment]
Thank you for reviewing this PR, @richm. Merging.

File: design_docs/logging_subsystem.md [hunk 17]
```diff
@@ -8,7 +8,7 @@ It is designed to satisfy following requirements.
 - The existing scenario, especially an input ovirt + output elasticsearch should not be affected by the updates.
 
 ```
-logging/
+logging
```
Comment (pcahyna) [inline design_docs/logging_subsystem.md:11]: Is it useful to have this long directory tree listing in this file that you must update every time you change the role structure? Maybe consider removing it in some future PR? @jvymazal @richm do you find useful to have this in the document?

Comment (jvymazal) [reply design_docs/logging_subsystem.md:11]: For me, probably not that much, if I need it, I can always browse it either here or locally.

Comment (richm) [reply design_docs/logging_subsystem.md:11]: maybe a description of the general layout, but not the specific files e.g. logging -> roles -> rsyslog -> tasks -> output -> task file for each output
